Sources/AsyncHTTPClient/HTTPHandler.swift

```diff
@@ -644,84 +639,34 @@ internal struct RedirectHandler<ResponseType> {
-        switch status {
-        case .movedPermanently, .found, .seeOther, .notModified, .useProxy, .temporaryRedirect, .permanentRedirect:
-            break
-        default:
-            return nil
-        }
-
-        guard let location = headers.first(name: "Location") else {
-            return nil
-        }
-
-        guard let url = URL(string: location, relativeTo: request.url) else {
-            return nil
-        }
-
-        guard self.request.deconstructedURL.scheme.supportsRedirects(to: url.scheme) else {
-            return nil
-        }
-
-        if url.isFileURL {
-            return nil
-        }
-
-        return url.absoluteURL
+    func redirectTarget(status: HTTPResponseStatus, responseHeaders: HTTPHeaders) -> URL? {
+        responseHeaders.extractRedirectTarget(
+            status: status,
+            originalURL: self.request.url,
+            originalScheme: self.request.deconstructedURL.scheme
+        )
     }
 
-    func redirect(status: HTTPResponseStatus, to redirectURL: URL, promise: EventLoopPromise<ResponseType>) {
-        var nextState: HTTPClient.Request.RedirectState?
-        if var state = request.redirectState {
-            guard state.count > 0 else {
-                return promise.fail(HTTPClientError.redirectLimitReached)
-            }
-
-            state.count -= 1
-
-            if var visited = state.visited {
-                guard !visited.contains(redirectURL) else {
-                    return promise.fail(HTTPClientError.redirectCycleDetected)
-                }
-
-                visited.insert(redirectURL)
-                state.visited = visited
-            }
-
-            nextState = state
-        }
-
-        let originalRequest = self.request
-
-        var convertToGet = false
-        if status == .seeOther, self.request.method != .HEAD {
-            convertToGet = true
-        } else if status == .movedPermanently || status == .found, self.request.method == .POST {
-            convertToGet = true
-        }
-
-        var method = originalRequest.method
-        var headers = originalRequest.headers
-        var body = originalRequest.body
-
-        if convertToGet {
-            method = .GET
-            body = nil
-            headers.remove(name: "Content-Length")
-            headers.remove(name: "Content-Type")
-        }
-
-        if !originalRequest.url.hasTheSameOrigin(as: redirectURL) {
-            headers.remove(name: "Origin")
-            headers.remove(name: "Cookie")
-            headers.remove(name: "Authorization")
-            headers.remove(name: "Proxy-Authorization")
-        }
-
+    func redirect(
+        status: HTTPResponseStatus,
+        to redirectURL: URL,
+        promise: EventLoopPromise<ResponseType>
+    ) {
         do {
+            var redirectState = self.request.redirectState
+            try redirectState?.redirect(to: redirectURL.absoluteString)
```

Comment:
Is it acceptable for `redirectState` to be `nil` here?

Reply:
No, I have removed `redirectState` from `Request` and moved it to `RedirectHandler` itself. We now explicitly forward the `RedirectState` and don't carry it implicitly through the Request.